File: Flutter-client/lib/common/api/bank_card/bank_card_api.dart
// ignore_for_file: unused_element

import '../base/base_api.dart';

enum BankCard with GoGamingApi {
  delete(data: {
    'id': 0,
    'key': 'x',
  }),
  batchDelete(data: {
    'ids': [0],
    'key': 'x',
  }),
  updateDefault(data: {
    'id': 0,
  }),
  getBankCard(),
  add(data: {
    'currencyType': 'Unknown',
    'name': 'string',
    'bankCode': 'string',
    'cardNum': 'string',
    'key': 'string',
  }),
  getSystemBank(params: {
    'currency': 'x',
  }),
  getDepositBankCard(params: {
    'currency': 'x',
    'paymentCode': 'x',
  }),
  verifyInfo(data: {
    'currencyType': 'x',
    'cardNum': 'x',
  });

  const BankCard({this.params, this.data});

  @override
  final Map<String, dynamic>? params;
  @override
  final Map<String, dynamic>? data;

  @override
  String get path {
    switch (this) {
      case BankCard.getBankCard:
        return '/asset/bankcard/getbankcard';
      case BankCard.delete:
        return '/asset/bankcard/deletebankcard';
      case BankCard.batchDelete:
        return '/asset/bankcard/batchdeletebankcard';
      case BankCard.updateDefault:
        return '/asset/bankcard/updatedefault';
      case BankCard.getSystemBank:
        return '/asset/bankcard/getsystembank';
      case BankCard.add:
        return '/asset/bankcard/addbankcard';
      case BankCard.getDepositBankCard:
        return '/asset/bankcard/getdepositbankcard';
      case BankCard.verifyInfo:
        return '/asset/bankcard/getbankcardverifyinfo';
    }
  }

  @override
  HTTPMethod get method {
    switch (this) {
      case BankCard.getBankCard:
      case BankCard.getSystemBank:
      case BankCard.getDepositBankCard:
        return HTTPMethod.get;
      case BankCard.delete:
      case BankCard.batchDelete:
      case BankCard.updateDefault:
      case BankCard.add:
      case BankCard.verifyInfo:
        return HTTPMethod.post;
    }
  }
}

File: Client-side/src/app/shared/apis/bank-card.api.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment';
import {
  AddBankCardPost,
  BankCard,
  BatchdeleteParam,
  DeleteParam,
  SystemBankCard,
  VerifyInfoParams,
  VerifyInfoResponse,
} from '../interfaces/bank-card.interface';
import { ResponseData } from '../interfaces/response.interface';
import { BaseApi } from './base.api';

@Injectable({
  providedIn: 'root',
})
export class BankCardApi extends BaseApi {
  private get url(): string {
    return `${environment.apiUrl}/v1/asset/bankcard`;
  }

  /**
   * 添加银行卡
   *
   * @param param
   * @returns
   */
  addBankCard(param: AddBankCardPost): Observable<ResponseData<boolean>> {
    param.cardNum = String(param.cardNum).replace(/ /g, '');
    return this.post(`${this.url}/addbankcard`, param);
  }

  /**
   * 获取银行卡
   *
   * @returns
   */
  getBankCard(): Observable<ResponseData<BankCard[]>> {
    return this.get(`${this.url}/getbankcard`);
  }

  /**
   * 设为预设银行卡
   *
   * @param cardId
   * @returns
   */
  setDefaultCard(cardId: number): Observable<ResponseData<boolean>> {
    const url = `${environment.apiUrl}/v1/asset/bankcard/updatedefault`;
    return this.post(url, { id: cardId });
  }

  /**
   * 获取系统支持银行资讯
   *
   * @param currency
   * @returns
   */
  getSystemBank(currency: string): Observable<ResponseData<SystemBankCard[]>> {
    return this.get(`${this.url}/getsystembank`, { currency: currency });
  }

  /**
   * 获取存款支付方式支持的银行卡
   *
   * @param paymentCode
   * @param currency
   * @returns
   */
  getDepositBank(paymentCode: string, currency: string): Observable<ResponseData<SystemBankCard[]>> {
    return this.get(`${this.url}/getdepositbankcard`, { paymentCode: paymentCode, currency: currency });
  }

  /**
   * 删除银行卡
   *
   * @param cardId
   * @param key
   * @param param
   * @returns
   */
  deleteCard(param: DeleteParam): Observable<ResponseData<boolean>> {
    return this.post(`${this.url}/deletebankcard`, param);
  }

  /**
   * 批量删除银行卡
   *
   * @param cardId
   * @param key
   * @param param
   * @returns
   */
  batchDeleteCard(param: BatchdeleteParam) {
    return this.post(`${this.url}/batchdeletebankcard`, param);
  }

  /**
   * 获取银行卡校验信息
   *
   * @param params currencyType, cardNum
   * @returns
   */
  getBankCardVerifyInfo(params: VerifyInfoParams): Observable<ResponseData<VerifyInfoResponse>> {
    params.cardNum = String(params.cardNum).replace(/ /g, '');
    return this.post(`${this.url}/getbankcardverifyinfo`, params);
  }
}
